src/profile-logic/transforms.js

```diff
@@ -906,70 +906,87 @@ export function collapseDirectRecursion(
-        !funcMatchesImplementation(thread, funcIndex))
-    ) {
-      // Out of N consecutive stacks that match the function to collapse, only remove
-      // stacks that are N > 1.
-      const newPrefixStackIndex = oldStackToNewStack.get(prefix);
-      if (newPrefixStackIndex === undefined) {
-        throw new Error('newPrefixStackIndex cannot be undefined');
+    const recursionChainPrefix = recursionChainPrefixForStack.get(prefix);
+    if (recursionChainPrefix === undefined) {
+      // Our prefix was not part of a recursion chain.
+      // If this stack frame matches the collapsed func, this stack node is the root
+      // of a recursion chain.
+      if (funcIndex === funcToCollapse) {
+        recursionChainPrefixForStack.set(stackIndex, prefix);
       }
-      oldStackToNewStack.set(stackIndex, newPrefixStackIndex);
-      recursiveStacks.add(stackIndex);
     } else {
-      // Add a stack in two cases:
-      //   1. It doesn't match the collapse requirements.
-      //   2. It is the first instance of a stack to collapse, re-use the stack and frame
-      //      information for the collapsed stack.
-      const newStackIndex = newStackTable.length++;
-      const newStackPrefix = oldStackToNewStack.get(prefix);
-      if (newStackPrefix === undefined) {
-        throw new Error(
-          'The newStackPrefix must exist because prefix < stackIndex as the StackTable is ordered.'
-        );
-      }
-      newStackTable.prefix[newStackIndex] = newStackPrefix;
-      newStackTable.frame[newStackIndex] = frameIndex;
-      newStackTable.category[newStackIndex] = category;
-      newStackTable.subcategory[newStackIndex] = subcategory;
-      oldStackToNewStack.set(stackIndex, newStackIndex);
-
-      if (funcToCollapse === funcIndex) {
-        recursiveStacks.add(stackIndex);
+      // Our prefix is part of a recursion chain.
+      if (funcMatchesImplementation(thread, funcIndex)) {
+        if (funcIndex === funcToCollapse) {
+          // The recursion chain continues.
+          recursionChainPrefixForStack.set(stackIndex, recursionChainPrefix);
+          // Reparent this stack node to the recursion root's prefix.
+          newStackTablePrefixColumn[stackIndex] = recursionChainPrefix;
+        } else {
+          // The recursion chain ends here. Leave recursionChainPrefixForStack
+          // empty for stackIndex.
+        }
+      } else {
+        // This stack node doesn't match the transform's implementation filter.
+        // For example, this stack node could be Xcpp in the following recursive
+        // JS invocation: Ajs -> Xcpp -> Ajs
+        // Keep the recursion chain going.
+        recursionChainPrefixForStack.set(stackIndex, recursionChainPrefix);
       }
     }
   }
-  return updateThreadStacks(
-    thread,
-    newStackTable,
-    getMapStackUpdater(oldStackToNewStack)
-  );
+
+  return {
+    ...thread,
+    stackTable: {
+      ...stackTable,
+      prefix: newStackTablePrefixColumn,
+    },
+  };
```

Comment:
Oh, I was confused at first that we might not be updating the `jsAllocations` and `nativeAllocations` tables.  But they will be updated still because we are updating the stackTable now directly instead of recreating the `samplesTable`. I guess this makes sense since we are always keeping the leaf-most stack and therefore we don't need to touch the `samplesTable`. Could you maybe add a comment explaining this?

Reply:
Sure, I'll add a comment.